natsyncd/natsync.cpp

```diff
@@ -0,0 +1,1047 @@
+    }
+}
+
+/* Conntrack notifications from the kernel don't have a flag to indicate if the
+ * NAT is NAPT or basic NAT. The original L4 port and the translated L4 port may 
+ * be the same and still can be the NAPT (can happen if the original L4 port is
+ * already in the L4 port pool range). To find out if it is a case of NAPT, we
+ * check if the Nat'ted IP is one of the NAPT pool range IP addresses, or if
+ * there is no static or dynamic NAT for that IP address. If both checks are not
+ * met, then it is the case of NAT basic translation where only IP address is NAT'ted. */
+bool NatSync::matchingSnaptPoolExists(const IpAddress &natIp)
+{
+    string key             = natIp.to_string();
+    std::vector<FieldValueTuple> values;
+
+    if (m_naptPoolCheckTable.get(key, values))
+    {
+        SWSS_LOG_INFO("Matching pool IP exists for NAT IP %s", key.c_str());
+        return true;
+    }
+
+    return false;
+}
+
+bool NatSync::matchingSnaptEntryExists(const naptEntry &entry)
+{
+    string key             = entry.orig_src_ip.to_string() + ":" + to_string(entry.orig_src_l4_port);
+    string reverseEntryKey = entry.nat_src_ip.to_string() + ":" + to_string(entry.nat_src_l4_port);
+    std::vector<FieldValueTuple> values;
+
+    if (m_naptCheckTable.get(key, values) || m_naptCheckTable.get(reverseEntryKey, values))
+    {
+        SWSS_LOG_INFO("Matching SNAPT entry exists for key %s or reverse key %s",
+                       key.c_str(), reverseEntryKey.c_str());
+        return true;
+    }
+    return false;
+}
+
+bool NatSync::matchingDnaptEntryExists(const naptEntry &entry)
+{
+    string key             = entry.orig_dest_ip.to_string() + ":" + to_string(entry.orig_dst_l4_port);
+    string reverseEntryKey = entry.nat_dest_ip.to_string() + ":" + to_string(entry.nat_dst_l4_port);
+    std::vector<FieldValueTuple> values;
+
+    if (m_naptCheckTable.get(key, values) || m_naptCheckTable.get(reverseEntryKey, values))
+    {
+        SWSS_LOG_INFO("Matching DNAPT entry exists for key %s or reverse key %s",
+                       key.c_str(), reverseEntryKey.c_str());
+        return true;
+    }
+    return false;
+}
+
+/* Add the NAT entries to APP_DB based on the criteria:
+ * ----------------------------------------------------
+ *   - If only source ip changed, add it as SNAT entry.
+ *   - If only destination ip changed, add it as DNAT entry.
+ *   - If SNAT happened and the l4 port changes or part of any dynamic pool range 
+ *       or if there is matching static or dynamic NAPT entry, add it as SNAPT entry.
+ *   - If DNAT happened and the l4 port changes or if there is matching static 
+ *       or dynamic NAPT entry, add it as DNAPT entry.
+ *   - If SNAT and DNAT happened, add it as Twice NAT entry.
+ *   - If SNAPT and DNAPT conditions are met or if there is no static
+ *       or dynamic Twice NAT entry, then it is a Twice NAPT entry.
+ */
+int NatSync::addNatEntry(struct nfnl_ct *ct, struct naptEntry &entry, bool addFlag)
```

Comment:
There are lot of DB gets and checks being done in this function. Will there be a performance impact in-terms of Netlink message loss, when the scale is high ?

Reply:
Per NAT entry, atmost 2 db gets are done and not seen netlink message loss with 600 pps traffic.
DB gets are done to avoid storing another copy of nat tables (apart from appl_db) within natsyncd.